grammars/prql-lezer/src/prql.grammar

```diff
@@ -3,114 +3,153 @@
-Ident { ident_part ( "." (ident_part | "*"))* }
+Ident { identPart ( "." (identPart | "*"))* }
+
+number { Integer | Float }
+
+kw<term> { @specialize[@name={term}]<identifier, term> }
 
-// I don't think it's possible to have `Op_bin` & `Op_unary` as tokens — that
-// would mean `-` can't be both unary & bin.
-Op_bin { op_bin_only | !term op_unary }
+VariableDeclaration { @specialize<identPart, "let"> identPart "=" (NestedPipeline (newline+ | end) | Lambda) }
 
-number { Integer time_unit? | Float }
+Lambda { LambdaParam* "->" exprCall ( newline+ | end ) }
+TypeDefinition { "<" TypeTerm ( "|" TypeTerm)* ">" }
+TypeTerm { identPart TypeDefinition? }
+LambdaParam { identPart TypeDefinition? (":" expression)? }
 
 @tokens {
+  CompareOp { "==" | "!=" | "~=" | ">=" | "<=" | ">" | "<" }
   date { @digit+ "-" @digit+ "-" @digit+ }
   time { @digit+ ":" @digit+ ( ":" @digit+ ( "." @digit+ )? )? }
   // We can't seem to set the number of digits, so this will allow any
   // combination of digits & hyphens.
   DateTime { "@" ( date | time | date "T" time ( "Z" | ( "-" | "+" ) @digit+ ":" @digit+ )? ) }
-  time_unit { "years" | "months" | "weeks" | "days" | "hours" | "minutes" | "seconds" | "milliseconds" | "microseconds" }
-  identifier_char { @asciiLetter | $[_\u{a1}-\u{10ffff}] }
-  ident_part { identifier_char (identifier_char | "_" | @digit )* }
+  TimeUnit { @digit+ ("years" | "months" | "weeks" | "days" | "hours" | "minutes" | "seconds" | "milliseconds" | "microseconds") }
+  identifierChar { @asciiLetter | $[_\u{a1}-\u{10ffff}] }
+  identPart { identifierChar (identifierChar | "_" | @digit )* }
+  identifier { identPart }
   Integer { @digit ( @digit | "_" )* ("e" Integer)? }
   Float { @digit ( @digit | "_" )* "." @digit ( @digit | "_" )* ("e" Integer)? }
   // TODO: This is not as precise as PRQL, which doesn't allow trailing
   // underscores and allows no digit before the decimal point.
   space { " "+ }
+  Docblock { "#!" ![\n]* }
   Comment { "#" ![\n]* }
-  op_bin_only { "*" | "/" | "//" | "%" | "!=" | ">=" | "<=" | "~=" | ">" | "<" | "??" | "&&" | "||" }
-  op_unary { "-" | "+" | "==" }
+  //op_bin_only { "*" | "/" | "//" | "%" | "!=" | ">=" | "<=" | "~=" | ">" | "<" | "??" | "&&" | "||" }
+  //op_unary { "-" | "+" | "==" }
   end { @eof }
-  line_wrap { "\\" }
-  wrapped_line { newline+ (Comment newline+)* line_wrap }
+  lineWrap { "\\" }
+  wrappedLine { newline+ (Comment newline+)* lineWrap }
   newline { "\n" }
   // TODO: Because this can also be used to compile to BETWEEN, ranges should
   // allow any literal, and arguably any expression.
-  Range { @digit+ ".." @digit+ }
+  RangeExpression { @digit+ ".." @digit+ }
   // Couldn't manage to do these & the interpolated as a template; couldn't
   // find how to negate a variable template
-  String { $["] !["]* $["] | $['] ![']* $['] }
+  String { '"""' ![Z]* '"""' | "'''" ![Z]* "'''" | $["] !["]* $["] | $['] ![']* $['] }
```

Comment:
What are the `Z`s for?

Reply:
It means any character except uppercase `Z`. It is a dirty workaround that I used because I don't know how to express any character except three consecutive double quotes.

While on the topic of strings, the [JavaScript grammar](https://github.com/lezer-parser/javascript/blob/main/src/javascript.grammar) have a `Escape` token. We would like that too. It can be mapped to [`escape`](https://lezer.codemirror.net/docs/ref/#highlight.tags.escape).

Reply:
OK, totally fine to have workarounds — but we have to have comments for these sorts of things, or the next person is going to have no idea why the `Z` is there!

I also don't completely understand how it's helping — why is it different from having "any character" — why does it treat `"` differently from "any character"? (quite possibly I'm misunderstanding...) 

Isn't it also going to break on any strict that contains `Z`?

Hope these questions are reasonable...

Reply:
You're right, I should have added a comment to make it more obvious. I have now added a comment, and also changed the `Z` to a 🍔 which is more obvious and less likely to be used.

I am not aware of any "any character" token in Lezer.

Yes, it is going to break on any string that contains `Z` (well now 🍔) but it's the price for the workaround, at least we got triple-quoted strings now.

Reply:
Thanks!

You're right about there not being anything like `@any`. I hadn't realized that didn't exist.

FYI this might be helpful: https://discuss.codemirror.net/t/matching-all-text-until-a-specific-token-is-encountered-without-externaltokenizer/5573/3, which links to https://lezer.codemirror.net/docs/guide/#local-token-groups